extensions/native/compiler/src/asm/compiler.rs

```diff
@@ -825,6 +835,103 @@ impl<F: PrimeField32 + TwoAdicField, EF: ExtensionField<F> + TwoAdicField> IfCom
     }
 }
 
+// Zipped for loop -- loop extends over the first entry in starts and ends
+pub struct ZipForCompiler<'a, F: Field, EF> {
+    compiler: &'a mut AsmCompiler<F, EF>,
+    starts: Vec<RVar<F>>,
+    ends: Vec<RVar<F>>,
+    step_sizes: Vec<F>,
+    loop_vars: Vec<Var<F>>,
+}
+
+impl<F: PrimeField32 + TwoAdicField, EF: ExtensionField<F> + TwoAdicField>
+    ZipForCompiler<'_, F, EF>
+{
+    pub(super) fn for_each(
+        self,
+        f: impl FnOnce(Vec<Var<F>>, &mut AsmCompiler<F, EF>),
+        debug_info: Option<DebugInfo>,
+    ) {
+        // initialize the loop variables
+        self.starts
+            .iter()
+            .zip(self.loop_vars.iter())
+            .for_each(|(start, loop_var)| match start {
+                RVar::Const(start) => {
+                    self.compiler.push(
+                        AsmInstruction::ImmF(loop_var.fp(), *start),
+                        debug_info.clone(),
+                    );
+                }
+                RVar::Val(start) => {
+                    self.compiler.push(
+                        AsmInstruction::CopyF(loop_var.fp(), start.fp()),
+                        debug_info.clone(),
+                    );
+                }
+            });
+
+        let loop_call_label = self.compiler.block_label();
+        let break_label = self.compiler.new_break_label();
+        self.compiler.break_label = Some(break_label);
+
+        self.compiler.basic_block();
+        let loop_label = self.compiler.block_label();
+
+        f(self.loop_vars.clone(), self.compiler);
+
+        self.loop_vars
+            .iter()
+            .zip(self.step_sizes.iter())
+            .for_each(|(loop_var, step_size)| {
+                self.compiler.push(
+                    AsmInstruction::AddFI(loop_var.fp(), loop_var.fp(), *step_size),
+                    debug_info.clone(),
+                );
+            });
+
+        self.compiler.basic_block();
+        let end = self.ends[0];
```

Comment:
We should document the behavior of this function, namely that it assumes that the first argument is the shortest and stops after iterating over the first argument.

Reply:
Related: Can we skip computing `ends[i]` for `i != 0`? It seems like we only need the final pointer for the first iterator.